### PR TITLE
Group API

### PR DESCRIPTION
**Pull request by aluzzardi**
The group API allows to programmatically group related operations together for status reporting and visualization purposes (e.g. logs).

Groups are hierarchical and can contain any number of sub-groups.

## TODO / Open Questions

- [x] Validate the API (**blocker**)
- [x] **codegen**: `client.Group()` currently returns a `*dagger.Query`
  - should codegen wrap `*dagger.Query` in a `*dagger.Client`, so it can be transparently passed around to unsuspecting `dagger.Client` receivers? (e.g. control grouping from the caller)
  - /cc @slumbering @dolanor @helderco 
- [ ] Q? Add `custom attributes` to groups/vertices?
  - Right now, it's just a `name` string
  - Built-in attributes (e.g. attached key/values) could provide a much better visualization experience (e.g. see the `ENV` for an `exec`)
- [ ] Group passthrough and events stream back for dagger-in-dagger
  - Currently, dagger-in-dagger is a black hole in terms of events
- [x] Bug in API stitching (**blocker**)
  - `extend type Query { group(...): Query! }` doesn't work (complains about `Query` being undefined
  - Sometimes this randomly fails (`group` field not found in `Query`)

## Example Usage

Taking this repository as an example, we perform hundreds of operations (e.g. exec, copies) which all get reported back as a single flat stream, with all operations mixed together.

This repository's CI is organized in 20 or so logical "parts": 3 SDKs with a build/lint/test/generate target each, an engine with a few targets, etc.

Using this API, each "logical part" would be grouped programmatically. The group hierarchy could look like this (simplified for illustration purposes):

| Group            | Description |
| ------------- | ----------- |
| `sdk`              | Parent group for all SDKs       |
| `sdk/go`         | Group for everything Go SDK related        |
| `sdk/go/lint`   | Sub-group for Go SDK linting        |
| `engine`          | Engine Group |
| `engine/build` | Group for engine build operations |

### Query API

- Groups can be created at the query level
- They are sticky (e.g. will affect the rest of the query)
- They are hierarchical (e.g. group can be called multiple times).

**Schema**

```graphql
extend type Query {
  group(name: String!): Query!
}
```

**Usage**
```go
// `c` belongs to the root group at this point
c, _ := dagger.Connect(ctx)

c.Group("a").Group("b").Group("c").Container(...).Exec(...)  // Exec is part of group `a/b/c`
```

**Example**
```go
// `c` belongs to the root group at this point
c, _ := dagger.Connect(ctx)

// everything using `c` is nested under the `sdk` group from now on
c = c.Group("sdk")

common := c.Container().From("alpine").Exec("codegen") // group is `sdk`

nodejs := c.Group("nodejs").Container().From("nodejs").Exec(...) // group is `sdk/nodejs`
golang := c.Group("golang").Container().From("golang").Exec(...) // group is `sdk/golang`
```

## Container/Directory API

- Groups can be created at the Container and Directory level
- They are nested under the parent Query group (or Container group)
- They are sticky (e.g. will affect the rest of the query)
- They are hierarchical (e.g. group can be called multiple times).

**API**
```graphql
extend type Container {
  group(name: String!): Container!
}

extend type Directory {
  group(name: String!): Directory!
}
```

**Usage**
```go
c, _ := dagger.Connect(ctx)

root := c.Group("a") // query-level group

alpine := root.Container().From("alpine").Exec("apk add bash") // nested under `a` (query-level group)

alpine.Group("build").Exec("sh build.sh") // group is `a/build`
alpine.Group("test").Exec("sh test.sh") // group is `a/test`
```

**Example**
```go
// `c` belongs to the root group at this point
c, _ := dagger.Connect(ctx)

// everything using `c` is nested under the `sdk` group from now on
c = c.Group("sdk")

nodejs := c.Group("nodejs").Container().From("nodejs").Exec("yarn install") // group is `sdk/nodejs`
lint := nodejs.Group("lint").Exec("yarn lint") // group is `sdk/nodejs/lint`
test := nodejs.Group("test").Exec("yarn test") // group is `sdk/nodejs/test`
```

## Built-in Groups

Some high-level operations in the API map to multiple low-level API calls, resulting in an incomprehensible mismatch. As in, executing a single operation can result in multiple "statuses" being sent back, without any way to correlate them.

The problem is even worse because of caching -- the number of low-level API calls might be different each run.

The API now creates a child-group automatically when performing multiple low-level operations, resulting in a 1:1 mapping for high-level calls to statuses.

Changes:

**c.Host().Directory()**

- **Before**
    - `local://<path>`
    - `copy / /`
- **With built-in group**
    -  `host.directory <path>`
        - `upload`
        - `copy`
    - **NOTE**: This one causes the most cache misses in our pipelines. Without grouping, it's tricky to debug because what matters is the cache status of `copy / /`, not the upload part. 

**c.Container().From()**

- **Before**
    - `resolve image config for docker.io/library/node:18-alpine`
    - `from alpine`
- **With built-in group**
    -  `from alpine`
        - `resolve image config`
        - `pull alpine`
    - **NOTE**: `resolve image config` shows up every time whereas the pulling part only if the image digest has changed from cache. Without sub-groups, every single `From()` maps back to 1 or 2 statuses, depending on cache. With sub-groups, there's always going to be a single `from` group for each `From()` call in a predictable way.

**c.Container().Build()**

- **Before**
    - Flat list of all Dockerfile steps, mixed together with your own operation logs
    - `[1/12] FROM docker.io/...`
    - `[2/12] RUN apk add curl`
- **With built-in group**
    - Every single operation resulting from docker build is nested together in a single group
    - `docker build`
        - `[1/12] FROM docker.io/...`
        - `[2/12] RUN apk add curl`
    - **NOTE**: Without this, currently all docker build logs are mixed together. As in, if we're building 4 docker files in a pipeline, we would see 4 times `[2/12] RUN apk add curl` with no idea which build it belongs to

/cc @sipsma @vito @shykes @helderco @dolanor

### CODE REVIEW

**Comment by shykes** [comment]
First: wow! This will take our DX *and* OX to the next level.

Second: my biggest question is on expected consumption model.
- I know this information will be invaluable for human operators, either in logs or other visualization UI in the console or web.
- Are there other ways that we anticipate this information being consumed? And does it involve introspecting groups from a graphql client? In other words is this groups API “write only” if that makes sense?

File: core/container.go
Comment (vito) [inline core/container.go:218]: Makes sense, just curious: does something (perhaps your CLI demo) parse this out? Don't see any other reference to `dagger.group` in this PR.

File: core/container.go
Comment (vito) [inline core/container.go:865]: Looks like this supports taking a progress group, so in this case the CustomName is just to make it clearer in the UI that this is an internal thing?

File: core/schema/group.go
Comment (vito) [inline core/schema/group.go:42]: Do we need to copy `parent`? :thinking:

**Review by vito — COMMENTED** [review]
API looks great! Just some minor comments.

File: core/container.go
Comment (sipsma) [reply core/container.go:218]: On this note, if it's straightforward I think we should update our current palin progress code to strip this out when present, otherwise the output looks like: `#3 [dagger.group:["from alpine:3.16.2"]] resolve image config for docker.io/library/alpine:3.16.2` which is kind of confusing.

File: router/schema.go
Comment (sipsma) [inline router/schema.go:116]: Think this is the source of the test errors, maybe can do
```go
if parent == nil { 
  parent = struct{}{}
}
```

**Comment by aluzzardi** [comment]
@sipsma @vito: Answering to both related comments, with a question:

> On this note, if it's straightforward I think we should update our current plain progress code to strip this out when present, otherwise the output looks like: #3 [dagger.group:["from alpine:3.16.2"]] resolve image config for docker.io/library/alpine:3.16.2 which is kind of confusing.

> Looks like this supports taking a progress group, so in this case the CustomName is just to make it clearer in the UI that this is an internal thing?

So there's 2 "special" naming conventions:

- `[dagger.group:["group", "sub"]] name`
  - Used for vertices that do not support progress groups (resolve image config)
- `[dagger.internal] name`
  - Used for "internal" stuff: right now it's just `mkdir /meta`
  - `mkdir /meta` shows up on every exec: it's noisy and pointless. It's doesn't map to something the user actually requested, doesn't produce any logs, doesn't take any time significant to compute.
  - UI just hides dagger.internal vertices. We could have a flag to optionally un-hide them

It's simple enough, however parsing this out is a bit hacky with edge cases, and we'd have to replicate the logic for every consumer.

What do you think of using JSON as vertex names? With `name`, `group` and `internal` keys:

```json
{"name": "go build"}
{"name": "docker.io/library/alpine:3.16.2", "group": ["from alpine:3.16.2"]}
{"name": "mkdir /meta", "internal": true}
```

In the future, we could actually make this fully structured:

```json
{"op": "exec", "args": ["go", "build"], "env": {"GOOS": "linux"}}
```

This would give a lot of liberty to the UI for rendering (e.g. color code file ops and execs differently) and observability (e.g. filter pipeline timers by execs, show me the slowest image pulls, ...)

File: core/group.go
Comment (aluzzardi) [reply core/group.go]: Fixed!

Also renamed `PG()` to `ProgressGroup()`

File: core/container.go
Comment (aluzzardi) [reply core/container.go]: Yes -- this appears as a random `mkdir /meta` and it's not obvious what it stands for (the user requested an exec and this randomly appears)

- `creating dagger metadata` is to make it more obvious than `mkdir /meta`
- `[dagger.internal]` is to hide it altogether from the UI

See comment below for an alternative

File: core/query.go
Comment (aluzzardi) [inline core/query.go:8]: side note: @sipsma if we ever want to, we could use this same mechanism for a `query.WithDefaultPlatform()`

**Comment by helderco** [comment]
> **codegen:** [...] should codegen wrap `*dagger.Query` in a `*dagger.Client`, [...] \cc @helderco 

In Python `dagger.Client` is actually `dagger.Query`:

https://github.com/dagger/dagger/blob/c5a9cdbdb9d01ad2c882463e4a539b8586f2eb1e/sdk/python/src/dagger/codegen.py#L495

**Comment by aluzzardi** [comment]
@vito @sipsma So there's a bug that prevents query fields from returning a query within an extend (e.g. `extend Query { field(...): Query`):

```
failed to resolve all type definitions: [Query]
```

Seems like it's a bug deep down in graphql-go-tools. I gave up debugging and instead added groups as part of the `query schema`, now in core (I removed the empty definition from the router).

Thoughts?

File: core/schema/container.graphqls
Comment (vito) [inline core/schema/container.graphqls:26]: Dupe?

```suggestion

```

Also needs a doc string

File: core/schema/directory.graphqls
Comment (vito) [inline core/schema/directory.graphqls:14]: needs a doc string

File: core/schema/query.graphqls
Comment (vito) [inline core/schema/query.graphqls:2]: doc string

**Review by vito — APPROVED** [review]
Sounds fine to me. :+1: 

The `.graphqls` just needs some tidying up, otherwise LGTM!

**Comment by dolanor** [comment]
My first reaction to this is just that `Group` might not be explicit enough.
`Container.Group` seems more of a group of containers than a group of event/log placed on this container instance.

I also don't like long names, but in the name of clarity, we could use: `ProgressGroup`, `LogGroup`, `EventGroup`? Open for suggestions.

**Comment by shykes** [comment]
As discussed live: I love the feature but not sure about 1) the name of the thing created ("group") and 2) the call to create it (`group`).

1: "group" is pretty abstract. It makes sense when you are in the right context. But from the user/developer point of view I worry there are many ways to interpret "group", for example is it a usergroup, a group of machines, etc. When we discussed this, the idea of "pipeline" or "sub-pipeline" or "nested pipeline" or "child pipeline" came up. Basically a pipeline within a pipeline. I like the consistency of it since we already talk about pipelines.

2: I like the simplicity of `withDescription`: attach a human-readable description to all operations beyond this point in the pipeline. Operations sharing the same description can be visualized as a virtual pipeline, with its own dependencies, runtime metrics etc. Later when/if we wish to attach more than one thing on the sub-pipeline, we can also offer `subPipeline` or just `pipeline` with an input object. But for now the convenience and familiarity of `withDescription` is IMO unmatched.

**Comment by shykes** [comment]
I also don't mind `pipeline`, but what's wrong with a long description? "Lint the Node.JS SDK using Node version 12" is exactly what the developer should be writing IMO.

How about `pipeline(name: String!, description: String): Query!`?

File: core/container.go
Comment (aluzzardi) [inline core/container.go:870]: @vito @sipsma Thoughts on this?

Basically serializing vertex names as JSON with `name` / `internal` / `group` (see previous comment)

Consumer is responsible for dealing with this. If the message fails JSON deserialization then assume it's `{name: <original payload>}`

**Comment by aluzzardi** [comment]
> I also don't mind pipeline, but what's wrong with a long description? 
> "Lint the Node.JS SDK using Node version 12" is exactly what the developer should be writing IMO.

To get the most of out it, I would programmatically split this in a few sub-pipelines (more or less, depending on the granularity):

`sdk` --> `node` --> `v12` --> `lint`

This way you can know how long the `node` SDK took. Was `lint` cached? From the UI I can collapse at the SDK level and see the list of languages.

This affects how things would be displayed in @vito's demo, and would be used as span service names for tracing (e.g. see tracing screenshot https://github.com/dagger/dagger/pull/4340#issuecomment-1376126255)

Using "Lint the Node.JS SDK using Node version 12" would useful to override `yarn lint` as a vertex name, but it's a different feature (i.e. akin to aliasing execs, rather than grouping a bunch of execs together)

File: core/container.go
Comment (vito) [reply core/container.go:870]: Definitely a bit hokey but I don't know a better alternative. Seems like something we could replace with upstream improvements in Buildkit (ability to marking things 'internal' or hidden by default, ability to set a ProgressGroup for resolve) -  but I wouldn't block on it.

Comment (aluzzardi) [reply core/container.go:870]: @vito Yep, indeed. BuildKit does support arbitrary key values (`WithCustomName` maps to `"llb.customname": name`), however those things don't make it back as solve status, AFAIK. Unless I'm mistaken?

**Comment by shykes** [comment]
> > I also don't mind pipeline, but what's wrong with a long description?
> > "Lint the Node.JS SDK using Node version 12" is exactly what the developer should be writing IMO.
> 
> To get the most of out it, I would programmatically split this in a few sub-pipelines (more or less, depending on the granularity):
> 
> `sdk` --> `node` --> `v12` --> `lint`
> 
> This way you can know how long the `node` SDK took. Was `lint` cached? From the UI I can collapse at the SDK level and see the list of languages.

Ok I see what you mean. I agree that human-readable descriptions should not be used *instead of* nesting. I was hoping they could be used *in addition* to it. So, in your example, perhaps each of `sdk`, `node`, `v12` and `lint` could have an (optional) description field? IMO it's also a good test: if you can't write even a succint description of what your pipeline does; that's probably a bad sign.

Re-iterating my new favorite option, how about:

```graphql
extend type Query {
  """
  Define a new pipeline, for logical grouping and observability.
  Pipelines can be nested. The top-level pipeline is implicit.
  """
  pipeline(name: String!, description: String): Query!
}
```

**Comment by aluzzardi** [comment]
> Re-iterating my new favorite option, how about:
> 
> ```graphql
> extend type Query {
>   """
>   Define a new pipeline, for logical grouping and observability.
>   Pipelines can be nested. The top-level pipeline is implicit.
>   """
>   pipeline(name: String!, description: String): Query!
> }
> ```

Sounds good to me! I'll have to rework the code a little bit to have multiple fields (name and description), I'll do that tomorrow.

I've taken care of the 2 other blockers (thanks @helderco and @dolanor!)  -- this PR should be in a mergeable state by tomorrow.

@vito @sipsma Let me know if you have any concerns about the implementation. Once merged I'll integrate into our own pipelines so we can start bubbling up the data

**Comment by dolanor** [comment]
I love where this is going! I like `pipeline`.

**Comment by vito** [comment]
Pipeline SGTM too! We're already calling the entire outer thing a "pipeline" colloquially, and this API call effectively forms a sub-pipeline or branch of the pipeline, so it feels consistent.

File: core/container.go
Comment (sipsma) [inline core/container.go:75]: extreme nit: "pipeline" instead of "group"

**Review by sipsma — APPROVED** [review]

File: core/container.go
Comment (aluzzardi) [reply core/container.go]: Thanks, yep, I missed a couple of those

**Comment by aluzzardi** [comment]
@sipsma re: your earlier comment:

> On this note, if it's straightforward I think we should update our current plain progress code to strip this out when present, otherwise the output looks like: #3 [dagger.group:["from alpine:3.16.2"]] resolve image config for docker.io/library/alpine:3.16.2 which is kind of confusing.

Just implemented that (engine.go). A bit nasty but does the trick (channel "proxying" for `DisplaySolveStatus`)

Dealing with one last test then this is good to go

**Comment by aluzzardi** [comment]
@vito I think I'm hitting the same issue you described here: https://github.com/dagger/dagger/issues/4130#issuecomment-1377578987

I've changed `query.directory` to do (simplified):

```go
	if args.ID != "" {
		return &core.Directory{
			ID: args.ID,
		}, nil
	}

	return core.NewDirectory(ctx, llb.Scratch(), "", pipeline, platform)
}
```

I believe the `core.NewDirectory()` with `llb.Scratch()` is causing a panic. Investigating

File: core/directory.go
Comment (aluzzardi) [inline core/directory.go:171]: @vito PTAL, I think this fixes #4130 (as well as this PR)

- Copied the original check logic for `llb.Scratch()` above, however check for `res.SingleRef()` returning `nil` (rather than checking `payload.LLB == nil`). AFAIK `gw.Solve() + SingleRef()` will always return `nil` for a `scratch` **result**
- Removed the original scratch check above for repetition. Solving `llb.Scratch()` (or `nil`) will result to a scratch result, so the edge case is already captured (the only downside is we lose a bit of time trying to Solve() scratch, but that's ok)
- Not sure what edge cases `path.Clean(src) == "."` is supposed to catch, but that didn't work for `Directory("")`. So I added an explicit `/` check to make this work for `scratch.Directory("")`, `scratch.Directory("/")`
- Added dir Mode to `Stat` of scratch. `container.Directory()` has a `if !info.IsDir() { return nil, fmt.Errorf("path %s is a file, not a directory", dirPath) }` check, which was failing

/cc @sipsma

Comment (vito) [reply core/directory.go:171]: Nice, `if ref == nil` feels a lot more sturdy. LGTM!

**Pull request merged**
